TESTS/mbed_platform/FileHandle/TestFile.h

```diff
@@ -0,0 +1,170 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_TESTFILEHANDLE_H
+#define MBED_TESTFILEHANDLE_H
+
+#include "platform/FileHandle.h"
+
+
+#define POS_IS_VALID(pos) (pos >= 0 && pos < _end)
+#define NEW_POS_IS_VALID(pos) (pos >= 0 && pos < (int32_t)FILE_SIZE)
+#define SEEK_POS_IS_VALID(pos) (pos >= 0 && pos <= _end)
+#define FILE_IS_EMPTY() (_end == 0)
+
+template<uint32_t FILE_SIZE>
+class TestFile : public FileHandle {
+public:
+    TestFile(): _pos(0), _end(0) {}
+    ~TestFile() {}
+
+    enum FunctionName {
+        fnNone, fnRead, fnWrite, fnSeek, fnClose, fnIsatty
+    };
+
+    virtual ssize_t read(void *buffer, size_t size)
+    {
+        ssize_t read;
+        _fnCalled = fnRead;
+
+        if(0 == size) {
+            return -EAGAIN;
+        }
+
+        for(read = 0; (size_t)read < size; read++)
+        {
+            if(POS_IS_VALID(_pos)) {
+                ((uint8_t*)buffer)[read] = _data[_pos++];
+            } else {
+                if(0 == read) {
+                    read = -EAGAIN;
+                }
+                break;
+            }
+        } // for
+        return read;
+    }
+
+    virtual ssize_t write(const void *buffer, size_t size)
+    {
+        ssize_t written;
+        _fnCalled = fnWrite;
+
+        if(0 == size) {
+            return -EAGAIN;
+        }
+
+        for(written = 0; (size_t)written < size; written++)
+        {
+            if(NEW_POS_IS_VALID(_pos)) {
+                _data[_pos++] = ((uint8_t*)buffer)[written];
+            } else {
+                if(0 == written) {
+                    written = -EAGAIN;
+                }
+                break;
+            }
+            if(_end < _pos) {
+                _end++;
+            }
+        } // for
+        return written;
+    }
+
+    virtual off_t seek(off_t offset, int whence)
+    {
+        _fnCalled = fnSeek;
+
+        switch(whence)
+        {
+            case SEEK_SET:
+            {
+                if(SEEK_POS_IS_VALID(offset)) {
+                    _pos = offset;
+                } else {
+                    return -1;
+                }
+                break;
+            }
+
+            case SEEK_CUR:
+            {
+                const int32_t new_pos = _pos + offset;
+                if(SEEK_POS_IS_VALID(new_pos)) {
```

Comment:
You're repeating yourself here - switch statement could just work out new_pos, leaving the "is valid" stuff common afterwards.